lib/ui/hash_codes.dart

```diff
@@ -28,95 +28,68 @@ int hashValues(
-                                result = 37 * result + arg17.hashCode;
-                                if (arg18 != _hashEnd) {
-                                  assert(arg18 is! Iterable);
-                                  result = 37 * result + arg18.hashCode;
-                                  if (arg19 != _hashEnd) {
-                                    assert(arg19 is! Iterable);
-                                    result = 37 * result + arg19.hashCode;
-                                    if (arg20 != _hashEnd) {
-                                      assert(arg20 is! Iterable);
-                                      result = 37 * result + arg20.hashCode;
-                                      // I can see my house from here!
-                                    }
-                                  }
-                                }
-                              }
-                            }
-                          }
-                        }
-                      }
-                    }
-                  }
-                }
-              }
-            }
-          }
-        }
-      }
-    }
+  void append(Object o) {
+    assert(o is! Iterable);
+    result = 0x1fffffff & (result + o.hashCode);
+    result = 0x1fffffff & (result + ((0x0007ffff & result) << 10));
+    result = result ^ (result >> 6);
   }
+  append(arg01);
+  append(arg02);
+  if (arg03 == _hashEnd) return result;
+  append(arg03);
+  if (arg04 == _hashEnd) return result;
+  append(arg04);
+  if (arg05 == _hashEnd) return result;
+  append(arg05);
+  if (arg06 == _hashEnd) return result;
+  append(arg06);
+  if (arg07 == _hashEnd) return result;
+  append(arg07);
+  if (arg08 == _hashEnd) return result;
+  append(arg08);
+  if (arg09 == _hashEnd) return result;
+  append(arg09);
+  if (arg10 == _hashEnd) return result;
+  append(arg10);
+  if (arg11 == _hashEnd) return result;
+  append(arg11);
+  if (arg12 == _hashEnd) return result;
+  append(arg12);
+  if (arg13 == _hashEnd) return result;
+  append(arg13);
+  if (arg14 == _hashEnd) return result;
+  append(arg14);
+  if (arg15 == _hashEnd) return result;
+  append(arg15);
+  if (arg16 == _hashEnd) return result;
+  append(arg16);
+  if (arg17 == _hashEnd) return result;
+  append(arg17);
+  if (arg18 == _hashEnd) return result;
+  append(arg18);
+  if (arg19 == _hashEnd) return result;
+  append(arg19);
+  if (arg20 == _hashEnd) return result;
+  append(arg20);
   return result;
 }
 
+
 /// Combine the hashCodes of an arbitrary number of values from an Iterable into
 /// one value. This function will return the same value if given "null" as if
 /// given an empty list.
 int hashList(Iterable<Object> args) {
   int result = 373;
+  void append(Object o) {
+    assert(o is! Iterable);
+    result = 0x1fffffff & (result + o.hashCode);
+    result = 0x1fffffff & (result + ((0x0007ffff & result) << 10));
+    result = result ^ (result >> 6);
+  }
```

Comment:
it might be cleaner to factor out `append` from both of these functions and have it take the current result and return the new result.